### PR TITLE
Raceconditions, SQL Support, SettingResolver, Error Handling...

### PR DESCRIPTION
**Pull request by kyranet**
### Proposed Semver Increment Bump: [MAJOR/MINOR/PATCH]
This PR was supposed to be Patch, fixing a racecondition, but ended fixing everything else. So... Minor?

### Changes Proposed in this Pull Request (List new items in CHANGELOG.MD)

- Fix ratecondition (providers now load before settingGateway).
- Better path resolving.
- SQL Compatibility.
- SQL property for Schema.
- getResolved method ( by @bdistin ) which returns the resolved configuration from a Guild.
- Better type parsing.
- settingResolver class.
- SQL class.
- `schema` getter for SettingGateway, as a shortcut of `SettingGateway.SchemaManager.schema`.
- `schemaManager` getter for Client, as a shortcut of `Client.SettingGateway.SchemaManager`.
- Better error catching at events.
- Added `SQL` and `settingResolver` to `index.js`.
- If the bot was unable to send a message, the `reboot` command would never call `process.exit()`.
- Beautified the `$conf list` command. Now it has the format from the `help` command.

### (If Applicable) What Issue does it fix?
 
- Raceconditions (Functions initing before Providers and SettingGateway).
- SQL Compatibility in the NoSQL Environment from SettingGateway.
- A few Unhandled Promise Errors.

### Testing

- [x] SQLite Compatibility.
- [x] SQL Schema Type Builder.
- [x] Conf command.
- [x] Array type (NoSQL).
- [x] Array type (SQL).

### CODE REVIEW

File: providers/json.js [hunk 1]
```diff
@@ -54,6 +55,16 @@ exports.getAll = (table) => {
 exports.get = (table, document) => fs.readJSON(resolve(baseDir, table, `${document}.json`)).catch(() => null);
 
 /**
+ * Check if the document exists.
+ * @param {string} table The name of the directory.
+ * @param {string} document The document name.
+ * @returns {Promise<boolean>}
+ */
+exports.has = (table, document) => fs.access(resolve(baseDir, table, `${document}.json`))
```
Comment (bdistin) [inline providers/json.js]: `exports.has = (table, document) => fs.pathExists(resolve(baseDir, table, `${document}.json`));`
pathExists is exactly written the way you have it there.

Comment (kyranet) [reply providers/json.js]: True, and returns boolean. I'll fix in a moment

Comment (kyranet) [reply providers/json.js]: Ao pls, it's literally the same thing. https://github.com/bdistin/fs-nextra/blob/master/index.js#L146

Comment (bdistin) [reply providers/json.js]: That's my point, lol! Why rewrite the function, when it already exists 😄

**Review by CyberiumShadow — APPROVED** [review]
LGTM

**Pull request merged**